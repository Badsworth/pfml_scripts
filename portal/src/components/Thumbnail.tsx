import React, { useEffect, useState } from "react";

/**
 * Create a URL for an image preview of image files and then trigger the component to render.
 * @param {File} file File whose preview we should load
 * @param {Function} setPreviewLoading State-setter to cause the component to render
 * @param {Function} setImageSrc State-setter to set the preview URL
 * @returns {undefined|Function} If a preview URL is loaded we'll return a function to clean up
 * the URL resource.
 */
const loadImagePreview = (file, setPreviewLoading, setImageSrc) => {
  if (file && file.type.startsWith("image/")) {
    const imageUrl = URL.createObjectURL(file);
    setImageSrc(imageUrl);
    setPreviewLoading(false);

    // On unmount we can clean up the URL.
    // Browsers will release object URLs automatically when the document is unloaded; however, for
    // optimal performance and memory usage, we can do it here manually when we know it's no longer
    // needed. Read more: https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL
    return () => URL.revokeObjectURL(imageUrl);
  } else {
    setPreviewLoading(false);
  }
};

interface ThumbnailProps {
  file?: any;
}

/**
 * Displays a thumbnail preview for a file input field
 */
const Thumbnail = (props: ThumbnailProps) => {
  const { file } = props;
  const [previewLoading, setPreviewLoading] = useState(true);
  const [imageSrc, setImageSrc] = useState();

  // Asynchronously load a preview of the file the first time this component is rendered.
  useEffect(() => {
    return loadImagePreview(file, setPreviewLoading, setImageSrc);
  }, [file]);

  if (previewLoading) {
    // Avoid showing a flash of the wrong content while we load a preview
    return null;
  }
  const hasImage = !!imageSrc;

  let thumbnailClass = "c-thumbnail margin-right-2 display-flex";
  if (!hasImage) {
    thumbnailClass =
      thumbnailClass + " c-thumbnail--fallback border-1px border-base-lighter";
  }

  return (
    <div className={thumbnailClass}>
      {hasImage ? (
        <img src={imageSrc} alt="" />
      ) : (
        <svg
          width="30"
          height="40"
          viewBox="0 0 30 40"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            stroke="black"
            strokeWidth="2"
            strokeLinecap="round"
            strokeLinejoin="round"
            d="M12 14.61H8M18.5 1.34H4.5C2.567 1.34 1 2.984 1 5.013V34.413C1 36.443 2.567 38.088 4.5 38.088H25.5C27.433 38.088 29 36.443 29 34.413V12.363L18.5 1.34ZM18 1.34V12.568H29L18 1.34ZM22 21.756H8H22ZM22 28.902H8H22Z"
          />
        </svg>
      )}
    </div>
  );
};

export default Thumbnail;
