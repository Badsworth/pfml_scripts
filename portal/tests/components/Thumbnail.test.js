/* eslint-disable import/first */
jest.mock("react", () => ({
  ...jest.requireActual("react"),
  // Mock useEffect so that we can manipulate it below
  useEffect: jest.fn(),
}));

import React, { useEffect } from "react";
import Thumbnail from "../../src/components/Thumbnail";
import { shallow } from "enzyme";

describe("Thumbnail", () => {
  const file = new File(["foo"], "foo.png", { type: "image/png" });

  beforeAll(() => {
    // URL.createObjectURL() hasn't been implemented in the jest DOM yet but will be
    // eventually. When it is (and this error triggers) we should remove this mock.
    // Read more: https://github.com/jsdom/jsdom/issues/1721
    if (URL.createObjectURL) {
      throw new Error(
        "jest DOM has added URL.createObjectURL() -- we can remove this hack now"
      );
    }
    URL.createObjectURL = () => "image.png";
    URL.revokeObjectURL = jest.fn();
  });

  describe("when initially loading the preview", () => {
    it("renders nothing", () => {
      const wrapper = shallow(<Thumbnail file={file} />);

      expect(wrapper.isEmptyRender()).toBe(true);
    });
  });

  describe("when selected file is a PDF", () => {
    it("renders placeholder thumbnail", () => {
      useEffect.mockImplementationOnce((didUpdate) => didUpdate());
      const pdfFile = new File(["foo"], "bar.pdf", { type: "application/pdf" });

      const wrapper = shallow(<Thumbnail file={pdfFile} />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when selected file is an image", () => {
    it("renders the img using a created URL", () => {
      useEffect.mockImplementationOnce((f) => f());
      const createUrlSpy = jest.spyOn(URL, "createObjectURL");

      const wrapper = shallow(<Thumbnail file={file} />);

      expect(createUrlSpy).toHaveBeenCalledTimes(1);
      expect(wrapper.prop("style")).toMatchInlineSnapshot(`
        Object {
          "backgroundImage": "url(image.png)",
        }
      `);
    });

    it("cleans up the created URL", () => {
      useEffect.mockImplementationOnce((f) => {
        const unmount = f();
        unmount();
      });
      const revokeUrlSpy = jest.spyOn(URL, "revokeObjectURL");

      shallow(<Thumbnail file={file} />);

      expect(revokeUrlSpy).toHaveBeenCalledTimes(1);
    });
  });
});
