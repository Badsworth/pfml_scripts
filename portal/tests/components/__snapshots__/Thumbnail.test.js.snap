// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thumbnail when selected file is a PDF renders placeholder thumbnail 1`] = `
<div
  class="c-thumbnail margin-right-2 display-flex c-thumbnail--fallback border-1px border-base-lighter"
>
  <svg
    fill="none"
    height="40"
    viewBox="0 0 30 40"
    width="30"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12 14.61H8M18.5 1.34H4.5C2.567 1.34 1 2.984 1 5.013V34.413C1 36.443 2.567 38.088 4.5 38.088H25.5C27.433 38.088 29 36.443 29 34.413V12.363L18.5 1.34ZM18 1.34V12.568H29L18 1.34ZM22 21.756H8H22ZM22 28.902H8H22Z"
      stroke="black"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
    />
  </svg>
</div>
`;
